### PR TITLE
Add CLI (argparse version).

### PR DESCRIPTION
**Pull request by grahamgower**
There are a few differences here compared with #338. In particular, the `ms` subcommand is a first class citizen with proper help output (reusing the existing ms parser argparse code, which is somewhere between complicated and impossible when using click). I don't have a strong opinion on what the "validate" subcommand is called, or how the options are spelled, but I think the options/defaults in this PR strike better balance than #338.

```
$ python -m demes -h
usage: __main__.py [-h] {validate,ms} ...

Validate and convert Demes models

positional arguments:
  {validate,ms}
    validate     Validate a model.
    ms           Build a demes model using ms commands.

optional arguments:
  -h, --help     show this help message and exit
```

```
$ python -m demes validate -h
usage: __main__.py validate [-h] [-j] [-s] [filename]

positional arguments:
  filename          Filename of the model. If not specified, the model will be read
                    from stdin.

optional arguments:
  -h, --help        show this help message and exit
  -j, --json        Output a JSON-formatted model. YAML is output by default.
  -s, --simplified  Output simplified model.
```

```
$ python -m demes ms -h
usage: __main__.py ms [-h] -N0 REFERENCE_SIZE [-f FILENAME] [-I value [value ...]]
                      [-n i x] [-g i α] [-G α] [-m i j rate] [-ma entry [entry ...]]
                      [-eG t α] [-eg t i α] [-eN t x] [-en t i x] [-eM t x]
                      [-em t i j rate] [-ema entry [entry ...]] [-es t i p]
                      [-ej t i j]

optional arguments:
  -h, --help            show this help message and exit
  -N0 REFERENCE_SIZE, --reference-size REFERENCE_SIZE
                        The reference population size used to translate from
                        coalescent units. This can be calculated from an existing ms
                        command as N0 = theta / (4 * μ * L), where μ is the per-
                        generation mutation rate and L is the length of the sequence
                        being simulated.
  -f FILENAME, --filename FILENAME
                        Insert commands from a file at this point in the command
                        line.
  -I value [value ...], --structure value [value ...]
                        Sample from populations with the specified deme structure.
                        The arguments are of the form 'num_demes n1 n2 ... [4N0m]',
                        specifying the number of populations, the sample
                        configuration, and optionally, the migration rate for a
                        symmetric island model
  -n i x, --population-size i x
                        Set the size of a specific population to x * N0.
  -g i α, --population-growth-rate i α
                        Set the growth rate to α for a specific population.
  -G α, --growth-rate α
                        Set the growth rate to α for all populations.
  -m i j rate, --migration-matrix-entry i j rate
                        Sets an entry M[i, j] in the migration matrix to the
                        specified rate. i and j are (1-indexed) population IDs.
                        Multiple options can be specified.
  -ma entry [entry ...], --migration-matrix entry [entry ...]
                        Sets the migration matrix to the specified value. The
                        entries are in the order M[1, 1], M[1, 2], ..., M[2, 1],
                        M[2, 2], ..., M[N, N], where N is the number of populations.
                        Diagonal entries may be written as 'x'.
  -eG t α, --growth-rate-change t α
                        Set the growth rate for all populations to α at time t
  -eg t i α, --population-growth-rate-change t i α
                        Set the growth rate for population i to α at time t
  -eN t x, --size-change t x
                        Set the population size for all populations to x * N0 at
                        time t
  -en t i x, --population-size-change t i x
                        Set the population size for population i to x * N0 at time t
  -eM t x, --migration-rate-change t x
                        Set the symmetric island model migration rate to x / (npop -
                        1) at time t
  -em t i j rate, --migration-matrix-entry-change t i j rate
                        Sets an entry M[i, j] in the migration matrix to the
                        specified rate at the specified time. i and j are
                        (1-indexed) population IDs.
  -ema entry [entry ...], --migration-matrix-change entry [entry ...]
                        Sets the migration matrix to the specified value at time t.
                        The entries are in the order M[1, 1], M[1, 2], ..., M[2, 1],
                        M[2, 2], ..., M[N, N], where N is the number of populations.
                        Diagonal entries may be written as 'x'.
  -es t i p, --admixture t i p
                        Split population i into a new population, such that the
                        specified proportion p of lineages remains in the population
                        i. Forwards in time this corresponds to an admixture event.
                        The new population has ID num_demes + 1. Migration rates to
                        and from the new population are set to 0, and growth rate is
                        0 and the population size for the new population is N0.
  -ej t i j, --population-split t i j
                        Move all lineages in population i to j at time t. Forwards
                        in time, this corresponds to a population split in which
                        lineages in j split into i. All migration rates for
                        population i are set to zero.
```

```
$ python -m demes ms -G 6.93 -eG 0.2 0.0 -eN 0.3 0.5 -N0 20000 | python -m demes validate -j
{
  "time_units": "generations",
  "demes": [
    {
      "name": "deme1",
      "start_time": "Infinity",
      "epochs": [
        {
          "end_time": 24000.0,
          "start_size": 10000.0,
          "end_size": 10000.0,
          "size_function": "constant",
          "selfing_rate": 0,
          "cloning_rate": 0
        },
        {
          "end_time": 16000.0,
          "start_size": 5001.472022241883,
          "end_size": 5001.472022241883,
          "size_function": "constant",
          "selfing_rate": 0,
          "cloning_rate": 0
        },
        {
          "end_time": 0,
          "start_size": 5001.472022241883,
          "end_size": 20000.0,
          "size_function": "exponential",
          "selfing_rate": 0,
          "cloning_rate": 0
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: demes/__main__.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    else:
+        output_format = "yaml"
+
+    multidoc, graphs = load_maybe_multidoc(args.filename)
+    if multidoc:
+        if output_format != "yaml":
+            raise RuntimeError(
+                "multi-document files only supported with --output-format yaml",
+            )
+            exit(1)
```
Comment (grahamgower) [inline demes/__main__.py:42]: Oops!
```suggestion
```

**Comment by jeromekelleher** [comment]
I don't have a strong opinion here - but either way it's a "nice to have"...

**Comment by jeromekelleher** [comment]
@grahamgower - looks good. I'm going to mute this thread though, so please ping me if you want some feedback.

**Comment by grahamgower** [comment]
@jeromekelleher If you have any ideas about the test failures on windows, I'd love to hear them!

**Comment by jeromekelleher** [comment]
Looks like a standard multiple file access problem on windows - it's a lot stricter about reusing file handles etc. For example, you can't reopen a NamedTemporaryFile on windows, even though it works fine on unix. Easiest to make a TemporaryDirectory in my experience, so you're sure the files are unique.

**Comment by grahamgower** [comment]
But @jeromekelleher, this *is* in a TemporaryDirectory...?

**Comment by jeromekelleher** [comment]
Ah, I see. Interesting...

**Comment by jeromekelleher** [comment]
Looks like the multidoc.yaml file is still open, when you try to delete the temp directory. Are you leaking a file handle in the CLI?

File: demes/__main__.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        self, filename: Union[str, io.TextIOBase]
+    ) -> Tuple[int, Iterator[demes.Graph]]:
+        """
+        Count the documents in the file, returning the count and an iterator
+        over the graphs. The returned document count is:
+            0 for zero documents,
+            1 for one document, and
+            2 for two or more documents.
+        """
+        graph_generator = demes.load_all(filename)
```
Comment (jeromekelleher) [inline demes/__main__.py:80]: Here maybe? Something to do with the generator not being unwound when the error is thrown? Seems unlikely tbh.

**Comment by jeromekelleher** [comment]
Not obvious to me what's going on - but I'll just reiterate the "is multidoc really worth it in the initial CLI" comment.

Other than that, well, you can always skip tests on windows. Simplest thing sometimes.

**Comment by grahamgower** [comment]
I reckon this is ready to merge now @jeromekelleher. I kept the multidoc bits in - its not so much extra code and I already wrote it. For the cli section in the docs, I tried [sphinx-argparse](https://github.com/alex-rudakov/sphinx-argparse), [sphinx-argparse-cli](https://github.com/gaborbernat/sphinx-argparse-cli) and  [sphinx-contrib.autoprogram](https://github.com/sphinx-contrib/autoprogram). None of them correctly show the option argument names/metavars, so I ended up just using the help output one sees at the terminal (using sphinxcontrib.programoutput), which I'd already tweaked to look sane.

File: tests/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+CONTINUOUS_INTEGRATION = os.getenv("CI") is not None
+
+
+class TestTopLevel:
+    def test_entrypoint1(self):
+        subprocess.run(
+            "python -m demes -h".split(), check=True, stdout=subprocess.DEVNULL
+        )
+
+    # If demes is "installed", then the "demes" command should be available.
```
Comment (jeromekelleher) [inline tests/test_cli.py:22]: Not sure about this bit - you're assuming that the module is always installed before we run tests in CI?

Could just test that the command ``demes`` works right after you've pip installed it in CI, and skip this.

**Review by jeromekelleher — APPROVED** [review]
LGTM. Haven't pored over the details, but looks very nice. One minor comment.

File: tests/test_cli.py (hunk 3 above)
Comment (grahamgower) [reply tests/test_cli.py]: Done. Thanks!

**Pull request merged**